test/common/upstream/health_check_fuzz.cc

```diff
@@ -0,0 +1,200 @@
+#include "test/common/upstream/health_check_fuzz.h"
+
+#include <chrono>
+
+#include "test/common/upstream/utility.h"
+#include "test/fuzz/utility.h"
+
+namespace Envoy {
+namespace Upstream {
+
+void HealthCheckFuzz::allocHealthCheckerFromProto(
+    const envoy::config::core::v3::HealthCheck& config) {
+  health_checker_ = std::make_shared<TestHttpHealthCheckerImpl>(
+      *cluster_, config, dispatcher_, runtime_, random_,
+      HealthCheckEventLoggerPtr(event_logger_storage_.release()));
+  ENVOY_LOG_MISC(trace, "Created Test Health Checker");
+}
+
+void HealthCheckFuzz::initializeAndReplay(test::common::upstream::HealthCheckTestCase input) {
+  try {
+    allocHealthCheckerFromProto(input.health_check_config());
+  } catch (EnvoyException& e) {
+    ENVOY_LOG_MISC(debug, "EnvoyException: {}", e.what());
+    return;
+  }
+  ON_CALL(runtime_.snapshot_, featureEnabled("health_check.verify_cluster", 100))
+      .WillByDefault(testing::Return(input.http_verify_cluster()));
+  cluster_->prioritySet().getMockHostSet(0)->hosts_ = {
+      makeTestHost(cluster_->info_, "tcp://127.0.0.1:80")};
+  expectSessionCreate();
+  expectStreamCreate(0);
+  if (input.start_failed()) {
+    cluster_->prioritySet().getMockHostSet(0)->hosts_[0]->healthFlagSet(
+        Host::HealthFlag::FAILED_ACTIVE_HC);
+  }
+  health_checker_->start();
+  ON_CALL(runtime_.snapshot_, getInteger("health_check.min_interval", _))
+      .WillByDefault(testing::Return(45000));
+  // If has an initial jitter, this calls onIntervalBase and finishes startup
+  if (input.health_check_config().initial_jitter().seconds() != 0) {
+    test_sessions_[0]->interval_timer_->invokeCallback();
+  }
+  reuse_connection_ = true;
+  if (input.health_check_config().has_reuse_connection()) { // TODO: Does this make sense?
+    reuse_connection_ = input.health_check_config().reuse_connection().value();
+  }
+  replay(input);
+}
+
+void HealthCheckFuzz::respondHttp(test::fuzz::Headers headers, absl::string_view status) {
+
+  // Timeout timer needs to be explicitly enabled, usually by onIntervalBase() (Callback on interval
+  // timer).
+  if (!test_sessions_[0]->timeout_timer_->enabled_) {
+    ENVOY_LOG_MISC(trace, "Timeout timer is disabled. Skipping response.");
+    return;
+  }
+
+  std::unique_ptr<Http::TestResponseHeaderMapImpl> response_headers =
+      std::make_unique<Http::TestResponseHeaderMapImpl>(
+          Fuzz::fromHeaders<Http::TestResponseHeaderMapImpl>(headers, {}, {}));
+
+  response_headers->setStatus(status);
+
+  ENVOY_LOG_MISC(trace, "Responded headers");
+
+  // Responding with http can cause client to close, if so create a new one.
+  bool client_will_close = false;
+  if (response_headers->Connection()) {
+    client_will_close =
+        absl::EqualsIgnoreCase(response_headers->Connection()->value().getStringView(),
+                               Http::Headers::get().ConnectionValues.Close);
+  }
+
+  if (response_headers->ProxyConnection()) {
```

Comment:
what if your headers are
Connection: Close
ProxyConnection: Garbage

Then client_will_close will be false because of the ProxyConnection case. Is that intended?
Edit: I think so, but only because codec client is HTTP/1. Might want to add a comment that right now this only fuzzes H/1

Reply:
Good catch! this is wrong.

Reply:
Took away that case.